// @cucumber/react/dist/src/humanFriendlyDuration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function preciseNumber(value, precision) {
    if (precision > 0) {
        return parseFloat(value.toFixed(precision));
    }
    return parseInt(value.toFixed(precision));
}
function splitFields(duration, readField, nextField, ratio, precision) {
    if (precision === void 0) { precision = 0; }
    var value = duration[readField];
    if (value >= ratio) {
        duration[nextField] = Math.trunc(value / ratio);
        duration[readField] = preciseNumber(value % ratio, precision);
    }
    return duration;
}
function hoursToDays(duration) {
    return splitFields(duration, 'hours', 'days', 24);
}
function minutesToHours(duration) {
    return splitFields(duration, 'minutes', 'hours', 60);
}
function secondsToMinutes(duration) {
    return splitFields(duration, 'seconds', 'minutes', 60, 3);
}
function humanFriendlyDuration(millis) {
    if (millis < 1000) {
        return { millis: preciseNumber(millis, 2) };
    }
    return hoursToDays(minutesToHours(secondsToMinutes({
        seconds: preciseNumber(millis / 1000, 3),
    })));
}
exports.default = humanFriendlyDuration;
