// @cucumber/react/dist/test/search/TagSearchTest.js
"use strict";
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var messages_1 = require("@cucumber/messages");
var assert_1 = __importDefault(require("assert"));
var gherkin_1 = require("@cucumber/gherkin");
var gherkin_utils_1 = require("@cucumber/gherkin-utils");
var TagSearch_1 = __importDefault(require("../../src/search/TagSearch"));
describe('TagSearchTest', function () {
    var gherkinQuery;
    var tagSearch;
    var feature = "@system\nFeature: Solar System\n\n  @planet\n  Scenario: Earth\n    Given is the sixth planet from the Sun\n\n  @dwarf\n  Scenario: Pluto\n    Given it is not really a planet\n  ";
    beforeEach(function () {
        gherkinQuery = new gherkin_utils_1.Query();
        tagSearch = new TagSearch_1.default(gherkinQuery);
    });
    function prettyResults(feature, query) {
        var e_1, _a, e_2, _b;
        var envelopes = gherkin_1.generateMessages(feature, 'test.feature', {
            includePickles: true,
            includeGherkinDocument: true,
            newId: messages_1.IdGenerator.incrementing(),
        });
        try {
            for (var envelopes_1 = __values(envelopes), envelopes_1_1 = envelopes_1.next(); !envelopes_1_1.done; envelopes_1_1 = envelopes_1.next()) {
                var envelope = envelopes_1_1.value;
                gherkinQuery.update(envelope);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (envelopes_1_1 && !envelopes_1_1.done && (_a = envelopes_1.return)) _a.call(envelopes_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        try {
            for (var envelopes_2 = __values(envelopes), envelopes_2_1 = envelopes_2.next(); !envelopes_2_1.done; envelopes_2_1 = envelopes_2.next()) {
                var envelope = envelopes_2_1.value;
                if (envelope.gherkinDocument) {
                    tagSearch.add(envelope.gherkinDocument);
                }
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (envelopes_2_1 && !envelopes_2_1.done && (_b = envelopes_2.return)) _b.call(envelopes_2);
            }
            finally { if (e_2) throw e_2.error; }
        }
        return gherkin_utils_1.pretty(tagSearch.search(query)[0]);
    }
    context('search', function () {
        it('returns an empty list when no documents have been added', function () {
            assert_1.default.deepStrictEqual(tagSearch.search('@any'), []);
        });
        it('finds matching scenarios', function () {
            assert_1.default.ok(prettyResults(feature, '@planet').includes('Scenario: Earth'));
        });
        it('takes into account feature tags', function () {
            var results = prettyResults(feature, '@system');
            assert_1.default.ok(results.includes('Scenario: Earth'));
            assert_1.default.ok(results.includes('Scenario: Pluto'));
        });
        it('supports complex search', function () {
            var results = prettyResults(feature, '@system and not @dwarf');
            assert_1.default.ok(results.includes('Scenario: Earth'));
            assert_1.default.ok(!results.includes('Scenario: Pluto'));
        });
    });
    context('with examples and tags', function () {
        // Currently, we are filtering at the scenario level,
        // so as long as one example match, we keep them all.
        var exampleFeature = "@system\nFeature: Solar system\n\n  Scenario: a planet may have sattelites\n    Then <satellites> should arround <planet>\n\n    @solid\n    Examples: solid planets\n      | planet | satellites     |\n      | earth  | moon           |\n      | mars   | phobos, demios |\n\n    @gas\n    Examples: giant gas planets\n      | jupiter | Io, Europe, Ganym\u00E8de, Callisto |\n";
        it('does not filter non-matching examples', function () {
            var results = prettyResults(exampleFeature, '@solid');
            assert_1.default.ok(results.includes('Scenario: a planet may have sattelites'));
            assert_1.default.ok(results.includes('Examples: solid planets'));
            assert_1.default.ok(results.includes('Examples: giant gas planets'));
        });
        it('does not filter examples which should be excluded', function () {
            var results = prettyResults(exampleFeature, '@solid and not @gas');
            assert_1.default.ok(results.includes('Scenario: a planet may have sattelites'));
            assert_1.default.ok(results.includes('Examples: solid planets'));
            assert_1.default.ok(results.includes('Examples: giant gas planets'));
        });
    });
});
