// @cucumber/react/dist/test/search/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeRule = exports.makeStep = exports.makeScenario = exports.makeFeature = void 0;
var messages_1 = require("@cucumber/messages");
function makeFeature(name, description, scenarios) {
    return messages_1.messages.GherkinDocument.Feature.create({
        name: name,
        description: description,
        children: scenarios.map(function (scenario) {
            return messages_1.messages.GherkinDocument.Feature.FeatureChild.create({
                scenario: scenario,
            });
        }),
    });
}
exports.makeFeature = makeFeature;
function makeRule(name, description, scenarios) {
    var idGenerator = messages_1.IdGenerator.uuid();
    return messages_1.messages.GherkinDocument.Feature.FeatureChild.Rule.create({
        id: idGenerator(),
        name: name,
        description: description,
        children: scenarios.map(function (scenario) {
            return messages_1.messages.GherkinDocument.Feature.FeatureChild.RuleChild.create({
                scenario: scenario,
            });
        }),
    });
}
exports.makeRule = makeRule;
function makeScenario(name, description, steps) {
    var idGenerator = messages_1.IdGenerator.uuid();
    return messages_1.messages.GherkinDocument.Feature.Scenario.create({
        id: idGenerator(),
        name: name,
        description: description,
        steps: steps,
    });
}
exports.makeScenario = makeScenario;
function makeStep(keyword, text, docstring, datatable) {
    if (docstring === void 0) { docstring = ''; }
    if (datatable === void 0) { datatable = []; }
    var idGenerator = messages_1.IdGenerator.uuid();
    var docStringParam = docstring
        ? messages_1.messages.GherkinDocument.Feature.Step.DocString.create({
            content: docstring,
        })
        : undefined;
    var datatableParam = datatable.length > 0
        ? messages_1.messages.GherkinDocument.Feature.Step.DataTable.create({
            rows: datatable.map(function (row) {
                return messages_1.messages.GherkinDocument.Feature.TableRow.create({
                    cells: row.map(function (cell) {
                        return messages_1.messages.GherkinDocument.Feature.TableRow.TableCell.create({
                            value: cell,
                        });
                    }),
                });
            }),
        })
        : undefined;
    return messages_1.messages.GherkinDocument.Feature.Step.create({
        id: idGenerator(),
        keyword: keyword,
        text: text,
        docString: docStringParam,
        dataTable: datatableParam,
    });
}
exports.makeStep = makeStep;
