// @wdio/cucumber-framework/build/reporter.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const cucumber_1 = require("@cucumber/cucumber");
const cucumberEventListener_1 = __importDefault(require("./cucumberEventListener"));
const utils_1 = require("./utils");
class CucumberReporter {
    constructor(eventBroadcaster, pickleFilter, _options, _cid, _specs, _reporter) {
        this._options = _options;
        this._cid = _cid;
        this._specs = _specs;
        this._reporter = _reporter;
        this.failedCount = 0;
        this._tagsInTitle = this._options.tagsInTitle || false;
        this._scenarioLevelReport = this._options.scenarioLevelReporter;
        this.eventListener = new cucumberEventListener_1.default(eventBroadcaster, pickleFilter)
            .on('before-feature', this.handleBeforeFeature.bind(this))
            .on('before-scenario', this.handleBeforeScenario.bind(this))
            .on('after-scenario', this.handleAfterScenario.bind(this))
            .on('after-feature', this.handleAfterFeature.bind(this));
        if (!this._scenarioLevelReport) {
            this.eventListener
                .on('before-step', this.handleBeforeStep.bind(this))
                .on('after-step', this.handleAfterStep.bind(this));
        }
    }
    handleBeforeFeature(uri, feature) {
        this._featureStart = new Date();
        this.emit('suite:start', {
            uid: utils_1.getFeatureId(uri, feature),
            title: this.getTitle(feature),
            type: 'feature',
            file: uri,
            tags: feature.tags,
            description: feature.description,
            keyword: feature.keyword
        });
    }
    handleBeforeScenario(uri, feature, scenario) {
        this._scenarioStart = new Date();
        this._testStart = new Date();
        this.emit(this._scenarioLevelReport ? 'test:start' : 'suite:start', {
            uid: scenario.id,
            title: this.getTitle(scenario),
            parent: utils_1.getFeatureId(uri, feature),
            type: 'scenario',
            file: uri,
            tags: scenario.tags
        });
    }
    handleBeforeStep(uri, feature, scenario, step) {
        this._testStart = new Date();
        const type = utils_1.getStepType(step);
        const payload = utils_1.buildStepPayload(uri, feature, scenario, step, { type });
        this.emit(`${type}:start`, payload);
    }
    handleAfterStep(uri, feature, scenario, step, result) {
        const type = utils_1.getStepType(step);
        if (type === 'hook') {
            return this.afterHook(uri, feature, scenario, step, result);
        }
        return this.afterTest(uri, feature, scenario, step, result);
    }
    afterHook(uri, feature, scenario, step, result) {
        var _a;
        let error;
        if (result.message) {
            error = new Error(result.message.split('\n')[0]);
            error.stack = result.message;
        }
        const payload = utils_1.buildStepPayload(uri, feature, scenario, step, {
            type: 'hook',
            state: result.status,
            error,
            duration: Date.now() - ((_a = this._testStart) === null || _a === void 0 ? void 0 : _a.getTime())
        });
        this.emit('hook:end', payload);
    }
    afterTest(uri, feature, scenario, step, result) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o;
        let state = 'undefined';
        switch (result.status) {
            case cucumber_1.Status.FAILED:
            case cucumber_1.Status.UNDEFINED:
                state = 'fail';
                break;
            case cucumber_1.Status.PASSED:
                state = 'pass';
                break;
            case cucumber_1.Status.PENDING:
            case cucumber_1.Status.SKIPPED:
            case cucumber_1.Status.AMBIGUOUS:
                state = 'pending';
        }
        let error = result.message ? new Error(result.message) : undefined;
        let title = step
            ? step === null || step === void 0 ? void 0 : step.text : this.getTitle(scenario);
        if (result.status === cucumber_1.Status.UNDEFINED) {
            if (this._options.ignoreUndefinedDefinitions) {
                /**
                 * mark test as pending
                 */
                state = 'pending';
                title += ' (undefined step)';
            }
            else {
                /**
                 * mark test as failed
                 */
                this.failedCount++;
                const err = new Error((step ? `Step "${title}" is not defined. ` : `Scenario ${title} has undefined steps. `) +
                    'You can ignore this error by setting cucumberOpts.ignoreUndefinedDefinitions as true.');
                err.stack = `${err.message}\n\tat Feature(${uri}):1:1\n`;
                const featChildren = (_a = feature.children) === null || _a === void 0 ? void 0 : _a.find(c => { var _a; return scenario.astNodeIds && ((_a = c.scenario) === null || _a === void 0 ? void 0 : _a.id) === scenario.astNodeIds[0]; });
                if (featChildren) {
                    err.stack += `\tat Scenario(${(_b = featChildren.scenario) === null || _b === void 0 ? void 0 : _b.name}):${(_d = (_c = featChildren.scenario) === null || _c === void 0 ? void 0 : _c.location) === null || _d === void 0 ? void 0 : _d.line}:${(_f = (_e = featChildren.scenario) === null || _e === void 0 ? void 0 : _e.location) === null || _f === void 0 ? void 0 : _f.column}\n`;
                    const featStep = (_h = (_g = featChildren.scenario) === null || _g === void 0 ? void 0 : _g.steps) === null || _h === void 0 ? void 0 : _h.find(s => step.astNodeIds && s.id === step.astNodeIds[0]);
                    if (featStep) {
                        err.stack += `\tat Step(${featStep.text}):${(_j = featStep.location) === null || _j === void 0 ? void 0 : _j.line}:${(_k = featStep.location) === null || _k === void 0 ? void 0 : _k.column}\n`;
                    }
                }
                error = err;
            }
        }
        else if (result.status === cucumber_1.Status.FAILED && !result.willBeRetried) {
            error = new Error((_l = result.message) === null || _l === void 0 ? void 0 : _l.split('\n')[0]);
            error.stack = result.message;
            this.failedCount++;
        }
        else if (result.status === cucumber_1.Status.AMBIGUOUS && this._options.failAmbiguousDefinitions) {
            state = 'fail';
            this.failedCount++;
            error = new Error((_m = result.message) === null || _m === void 0 ? void 0 : _m.split('\n')[0]);
            error.stack = result.message;
        }
        else if (result.willBeRetried) {
            state = 'retry';
        }
        const common = {
            title: title,
            state,
            error,
            duration: Date.now() - ((_o = this._testStart) === null || _o === void 0 ? void 0 : _o.getTime()),
            passed: state === 'pass',
            file: uri
        };
        const payload = step
            ? utils_1.buildStepPayload(uri, feature, scenario, step, {
                type: 'step',
                ...common
            })
            : {
                type: 'scenario',
                uid: scenario.id,
                parent: utils_1.getFeatureId(uri, feature),
                tags: scenario.tags,
                ...common
            };
        this.emit('test:' + state, payload);
    }
    handleAfterScenario(uri, feature, scenario, result) {
        var _a;
        if (this._scenarioLevelReport) {
            return this.afterTest(uri, feature, scenario, { id: scenario.id }, result);
        }
        this.emit('suite:end', {
            uid: scenario.id,
            title: this.getTitle(scenario),
            parent: utils_1.getFeatureId(uri, feature),
            type: 'scenario',
            file: uri,
            duration: Date.now() - ((_a = this._scenarioStart) === null || _a === void 0 ? void 0 : _a.getTime()),
            tags: scenario.tags
        });
    }
    handleAfterFeature(uri, feature) {
        var _a;
        this.emit('suite:end', {
            uid: utils_1.getFeatureId(uri, feature),
            title: this.getTitle(feature),
            type: 'feature',
            file: uri,
            duration: Date.now() - ((_a = this._featureStart) === null || _a === void 0 ? void 0 : _a.getTime()),
            tags: feature.tags
        });
    }
    emit(event, payload) {
        let message = utils_1.formatMessage({ payload });
        message.cid = this._cid;
        message.specs = this._specs;
        message.uid = payload.uid;
        this._reporter.emit(event, message);
    }
    getTitle(featureOrScenario) {
        const name = featureOrScenario.name;
        const tags = featureOrScenario.tags;
        if (!this._tagsInTitle || !tags || !tags.length)
            return name;
        return `${tags.map(tag => tag.name).join(', ')}: ${name}`;
    }
}
exports.default = CucumberReporter;
