// @cucumber/react/dist/src/search/FeatureSearch.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var elasticlunr_1 = __importDefault(require("elasticlunr"));
var FeatureSearch = /** @class */ (function () {
    function FeatureSearch() {
        this.featuresByUri = new Map();
        this.index = elasticlunr_1.default(function (ctx) {
            ctx.setRef('uri');
            ctx.addField('name');
            ctx.addField('description');
            ctx.saveDocument(true);
        });
    }
    FeatureSearch.prototype.add = function (gherkinDocument) {
        this.featuresByUri.set(gherkinDocument.uri, gherkinDocument.feature);
        this.index.addDoc({
            uri: gherkinDocument.uri,
            name: gherkinDocument.feature.name,
            description: gherkinDocument.feature.description,
        });
    };
    FeatureSearch.prototype.search = function (query) {
        var _this = this;
        var searchResultsList = this.index.search(query, {
            fields: {
                name: { bool: 'OR', boost: 1 },
                description: { bool: 'OR', boost: 1 },
            },
        });
        return searchResultsList.map(function (searchResults) {
            return _this.featuresByUri.get(searchResults.ref);
        });
    };
    return FeatureSearch;
}());
exports.default = FeatureSearch;
