// @cucumber/react/dist/src/search/RuleSearch.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var elasticlunr_1 = __importDefault(require("elasticlunr"));
var RuleSearch = /** @class */ (function () {
    function RuleSearch() {
        this.index = elasticlunr_1.default(function (ctx) {
            ctx.setRef('id');
            ctx.addField('name');
            ctx.addField('description');
            ctx.saveDocument(true);
        });
        this.ruleById = new Map();
    }
    RuleSearch.prototype.add = function (rule) {
        this.index.addDoc({
            id: rule.id,
            name: rule.name,
            description: rule.description,
        });
        this.ruleById.set(rule.id, rule);
    };
    RuleSearch.prototype.search = function (query) {
        var _this = this;
        var results = this.index.search(query, {
            fields: {
                name: { bool: 'OR', boost: 1 },
                description: { bool: 'OR', boost: 1 },
            },
        });
        return results.map(function (result) { return _this.ruleById.get(result.ref); });
    };
    return RuleSearch;
}());
exports.default = RuleSearch;
