// @cucumber/react/dist/src/search/ScenarioSearch.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var elasticlunr_1 = __importDefault(require("elasticlunr"));
var ScenarioSearch = /** @class */ (function () {
    function ScenarioSearch() {
        this.index = elasticlunr_1.default(function (ctx) {
            ctx.setRef('id');
            ctx.addField('name');
            ctx.addField('description');
            ctx.saveDocument(true);
        });
        this.scenarioById = new Map();
    }
    ScenarioSearch.prototype.add = function (scenario) {
        this.index.addDoc({
            id: scenario.id,
            name: scenario.name,
            description: scenario.description,
        });
        this.scenarioById.set(scenario.id, scenario);
    };
    ScenarioSearch.prototype.search = function (query) {
        var _this = this;
        var results = this.index.search(query, {
            fields: {
                name: { bool: 'OR', boost: 1 },
                description: { bool: 'OR', boost: 1 },
            },
        });
        return results.map(function (result) { return _this.scenarioById.get(result.ref); });
    };
    return ScenarioSearch;
}());
exports.default = ScenarioSearch;
