// @cucumber/react/dist/src/search/Search.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var TagSearch_1 = __importDefault(require("../../src/search/TagSearch"));
var TextSearch_1 = __importDefault(require("../../src/search/TextSearch"));
var is_tag_expression_1 = __importDefault(require("../../src/is_tag_expression"));
var Search = /** @class */ (function () {
    function Search(gherkinQuery) {
        this.gherkinQuery = gherkinQuery;
        this.textSearch = new TextSearch_1.default();
        this.tagSearch = new TagSearch_1.default(gherkinQuery);
    }
    Search.prototype.search = function (query) {
        if (is_tag_expression_1.default(query)) {
            try {
                return this.tagSearch.search(query);
            }
            catch (_a) {
                // No-op, we fall back to text search.
            }
        }
        return this.textSearch.search(query);
    };
    Search.prototype.add = function (gherkinDocument) {
        this.tagSearch.add(gherkinDocument);
        this.textSearch.add(gherkinDocument);
    };
    return Search;
}());
exports.default = Search;
