// @cucumber/react/dist/src/search/StepSearch.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var elasticlunr_1 = __importDefault(require("elasticlunr"));
var StepSearch = /** @class */ (function () {
    function StepSearch() {
        this.index = elasticlunr_1.default(function (ctx) {
            ctx.addField('keyword');
            ctx.addField('text');
            ctx.addField('docString');
            ctx.addField('dataTable');
            ctx.setRef('id');
            ctx.saveDocument(true);
        });
        this.stepById = new Map();
    }
    StepSearch.prototype.add = function (step) {
        var doc = {
            id: step.id,
            keyword: step.keyword,
            text: step.text,
            docString: this.docStringToString(step),
            dataTable: this.dataTableToString(step),
        };
        this.index.addDoc(doc);
        this.stepById.set(step.id, step);
    };
    StepSearch.prototype.search = function (query) {
        var _this = this;
        var results = this.index.search(query, {
            fields: {
                keyword: { bool: 'OR', boost: 1 },
                text: { bool: 'OR', boost: 2 },
                docString: { bool: 'OR', boost: 1 },
                dataTable: { bool: 'OR', boost: 1 },
            },
        });
        return results.map(function (result) { return _this.stepById.get(result.ref); });
    };
    StepSearch.prototype.docStringToString = function (step) {
        return step.docString ? step.docString.content : '';
    };
    StepSearch.prototype.dataTableToString = function (step) {
        return step.dataTable
            ? step.dataTable.rows
                .map(function (row) { return row.cells.map(function (cell) { return cell.value; }).join(' '); })
                .join(' ')
            : undefined;
    };
    return StepSearch;
}());
exports.default = StepSearch;
