// @cucumber/react/dist/src/search/TextSearch.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var FeatureSearch_1 = __importDefault(require("./FeatureSearch"));
var ScenarioSearch_1 = __importDefault(require("./ScenarioSearch"));
var StepSearch_1 = __importDefault(require("./StepSearch"));
var RuleSearch_1 = __importDefault(require("./RuleSearch"));
var gherkin_utils_1 = require("@cucumber/gherkin-utils");
var TextSearch = /** @class */ (function () {
    function TextSearch() {
        this.featureSearch = new FeatureSearch_1.default();
        this.backgroundSearch = new ScenarioSearch_1.default();
        this.scenarioSearch = new ScenarioSearch_1.default();
        this.stepSearch = new StepSearch_1.default();
        this.ruleSearch = new RuleSearch_1.default();
        this.gherkinDocuments = [];
    }
    TextSearch.prototype.search = function (query) {
        var matchingSteps = this.stepSearch.search(query);
        var matchingBackgrounds = this.backgroundSearch.search(query);
        var matchingScenarios = this.scenarioSearch.search(query);
        var matchingRules = this.ruleSearch.search(query);
        var matchingFeatures = this.featureSearch.search(query);
        var walker = new gherkin_utils_1.GherkinDocumentWalker({
            acceptStep: function (step) { return matchingSteps.includes(step); },
            acceptScenario: function (scenario) { return matchingScenarios.includes(scenario); },
            acceptBackground: function (background) {
                return matchingBackgrounds.includes(background);
            },
            acceptRule: function (rule) { return matchingRules.includes(rule); },
            acceptFeature: function (feature) { return matchingFeatures.includes(feature); },
        });
        return this.gherkinDocuments
            .map(function (gherkinDocument) { return walker.walkGherkinDocument(gherkinDocument); })
            .filter(function (gherkinDocument) { return gherkinDocument !== null; });
    };
    TextSearch.prototype.add = function (gherkinDocument) {
        var _this = this;
        this.gherkinDocuments.push(gherkinDocument);
        var walker = new gherkin_utils_1.GherkinDocumentWalker({}, {
            handleStep: function (step) { return _this.stepSearch.add(step); },
            handleScenario: function (scenario) { return _this.scenarioSearch.add(scenario); },
            handleBackground: function (background) { return _this.backgroundSearch.add(background); },
            handleRule: function (rule) { return _this.ruleSearch.add(rule); },
        });
        this.featureSearch.add(gherkinDocument);
        walker.walkGherkinDocument(gherkinDocument);
    };
    return TextSearch;
}());
exports.default = TextSearch;
