// @cucumber/react/dist/test/StepTest.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var assert_1 = __importDefault(require("assert"));
var react_dom_1 = __importDefault(require("react-dom"));
var react_1 = __importDefault(require("react"));
var messages_1 = require("@cucumber/messages");
var query_1 = require("@cucumber/query");
var Step_1 = __importDefault(require("../src/components/gherkin/Step"));
var jsdom_1 = require("jsdom");
var UriContext_1 = __importDefault(require("../src/UriContext"));
var GherkinQueryContext_1 = __importDefault(require("../src/GherkinQueryContext"));
var gherkin_utils_1 = require("@cucumber/gherkin-utils");
var CucumberQueryContext_1 = __importDefault(require("../src/CucumberQueryContext"));
describe('<Step>', function () {
    it('renders', function () {
        var dom = new jsdom_1.JSDOM('<html lang="en"><body><div id="content"></div></body></html>');
        // @ts-ignore
        global.window = dom.window;
        // global.navigator = dom.window.navigator
        var document = dom.window.document;
        var step = new messages_1.messages.GherkinDocument.Feature.Step({
            keyword: 'Given',
            text: 'the 48 pixies',
            location: new messages_1.messages.Location({ column: 1, line: 1 }),
        });
        var StubCucumberQuery = /** @class */ (function (_super) {
            __extends(StubCucumberQuery, _super);
            function StubCucumberQuery() {
                return _super !== null && _super.apply(this, arguments) || this;
            }
            StubCucumberQuery.prototype.getStepMatchArgumentsLists = function () {
                return [
                    new messages_1.messages.TestCase.TestStep.StepMatchArgumentsList({
                        stepMatchArguments: [
                            new messages_1.messages.TestCase.TestStep.StepMatchArgumentsList.StepMatchArgument({
                                group: new messages_1.messages.TestCase.TestStep.StepMatchArgumentsList.StepMatchArgument.Group({
                                    start: 4,
                                    value: '48',
                                    children: [],
                                }),
                            }),
                        ],
                    }),
                ];
            };
            return StubCucumberQuery;
        }(query_1.Query));
        var StubGherkinQuery = /** @class */ (function (_super) {
            __extends(StubGherkinQuery, _super);
            function StubGherkinQuery() {
                return _super !== null && _super.apply(this, arguments) || this;
            }
            StubGherkinQuery.prototype.getPickleStepIds = function () {
                return ['dummy-id'];
            };
            return StubGherkinQuery;
        }(gherkin_utils_1.Query));
        var app = (react_1.default.createElement(GherkinQueryContext_1.default.Provider, { value: new StubGherkinQuery() },
            react_1.default.createElement(UriContext_1.default.Provider, { value: 'some.feature' },
                react_1.default.createElement(CucumberQueryContext_1.default.Provider, { value: new StubCucumberQuery() },
                    react_1.default.createElement(Step_1.default, { step: step, renderStepMatchArguments: true, renderMessage: true })))));
        react_dom_1.default.render(app, document.getElementById('content'));
        var plainTexts = Array.from(document.querySelectorAll('#content h3 .cucumber-step__text, #content h3 .cucumber-step__keyword')).map(function (span) { return span.textContent; });
        assert_1.default.deepStrictEqual(plainTexts, ['Given', 'the ', ' pixies']);
        var paramTexts = Array.from(document.querySelectorAll('#content h3 .cucumber-step__param')).map(function (a) { return a.textContent; });
        assert_1.default.deepStrictEqual(paramTexts, ['48']);
    });
});
